File: Java/src/AOA/Sort/QuickSort.java
package AOA.Sort;

import java.io.*;
import java.util.Arrays;

public class QuickSort {
    static void swap(int arr[], int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    static int partition(int[] arr, int low, int high) {
        int pivot = arr[low];
        int temp;
        int l = low, h = high;
        while (l < h) {
            while (arr[l] <= pivot && l < h) l++;
            while (arr[h] > pivot) h--;
            if (l < h) {
                swap(arr, l, h);
            }
        }
        swap(arr, low, h);
        return h;
    }

    static void sort(int[] arr, int low, int high){
        if (low >= high)
            return;
        int piv = partition(arr, low, high);
        sort(arr,low, piv);
        sort(arr,piv+1,high);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        sort(arr, 0, arr.length - 1);
        for (int ar : arr) System.out.print(ar + " ");
    }

}

File: Python/AOA/sort/quick_sort.py
n = [4, 3, 2, 1]


def quick_sort(nums: list[int], s, e):
    if (e - s + 1 <= 1):
        return nums

    left, pivot = s,  e

    for i in range(s, e):
        if nums[i] < nums[pivot]:
            nums[i], nums[left] = nums[left], nums[i]
            left += 1

    nums[pivot], nums[left] = nums[left], nums[pivot]

    quick_sort(nums, s, left - 1)
    quick_sort(nums, left + 1, pivot)

    return nums


print(quick_sort(n, 0, len(n)-1))
